### PR TITLE
feature: allow multithreading code, run in Docker, precompile v fmt for Docker

### PR DESCRIPTION
**Pull request by walkingdevel**
before

-->

after

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ version: "3.9"
 services:
   playground:
     command: bash -c "cd /usr/src/app && make run"
+    privileged: true
```
Comment (spytheman) [inline docker-compose.yml:6]: why is this needed?

Comment (walkingdevel) [reply docker-compose.yml:6]: Isolate doesn't work without privileged mode in Docker.

File: server.v [hunk 2]
```diff
@@ -8,10 +8,17 @@ import sqlite
...
+	port                          = 5555
+	vexeroot                      = @VEXEROOT
+	// non-standard block size, different for different filesystems
+	block_size                    = 4096
+	fs_usage_max_size_in_bytes    = 10 * 1024 * 1024
+	// from isolate docs: please note that this currently works only on the ext family of filesystems (other filesystems use other interfaces for setting quotas)
+	block_max_count               = u32(fs_usage_max_size_in_bytes / block_size)
+	inode_max_count               = 50
+	max_run_processes_and_threads = 10
+	max_run_memory_in_kb          = 50_000
```
Comment (spytheman) [inline server.v:20]: This may not be enough for slightly larger programs, it would have to be tested.
Imho 100MB was a better limit.

File: docker-compose.yml (hunk 1 above)
Comment (walkingdevel) [reply docker-compose.yml:6]: Isolate throws "Cannot run proxy, clone failed: Operation not permitted".

File: server.v (hunk 2 above)
Comment (walkingdevel) [reply server.v:20]: Agree with you. I'll fix it.

Comment (spytheman) [reply server.v:20]: For example for me on Linux:
```
#0 20:51:17 ᛋ const-precision /v/vnew❱cp vlib/rand/dist_test.v x.v
#0 20:51:19 ᛋ const-precision /v/vnew❱xtime v x.v 
CPU: 0.38s      Real: 0.34s     Elapsed: 0:00.34        RAM: 54308KB    v x.v
```

File: server.v [hunk 3]
```diff
@@ -81,28 +90,28 @@ fn ddhhmmss(time time.Time) string {
...
+	box_path, box_id := try_init_sandbox()
 	defer {
-		isolate_cmd('isolate --box-id=$box_id --cleanup')
+		isolate_cmd('isolate --box-id=${box_id} --cleanup')
 	}
 	os.write_file(os.join_path(box_path, 'code.v'), code) or {
 		return 'Failed to write code to sandbox.'
 	}
-	build_res := isolate_cmd('isolate --box-id=$box_id --dir=$vexeroot --env=HOME=/box --processes=3 --mem=100000 --wall-time=2 --quota=${1048576 / block_size},${1048576 / inode_ratio} --run -- $vexeroot/v -cflags -DGC_MARKERS=1 -no-parallel -no-retry-compilation -g code.v')
+	build_res := isolate_cmd('isolate --box-id=${box_id} --dir=${vexeroot} --env=HOME=/box --processes=${max_run_processes_and_threads} --mem=${max_run_memory_in_kb} --wall-time=${run_time_in_seconds} --run -- ${vexeroot}/v -cflags -DGC_MARKERS=1 -no-parallel -no-retry-compilation -g code.v')
```
Comment (spytheman) [inline server.v:108]: Why is `--quota=` not passed anymore?
Is setting it in the box initialization enough?

Comment (walkingdevel) [reply server.v:108]: From docs: Set disk quota to a given number of blocks and inodes. This requires the filesystem to be mounted with support for quotas. Unlike other options, this one must be given to isolate --init.

File: server.v [hunk 4]
```diff
@@ -81,28 +90,28 @@ fn ddhhmmss(time time.Time) string {
...
-	box_path, box_id := init_sandbox()
+	box_path, box_id := try_init_sandbox()
 	defer {
-		isolate_cmd('isolate --box-id=$box_id --cleanup')
+		isolate_cmd('isolate --box-id=${box_id} --cleanup')
 	}
 	os.write_file(os.join_path(box_path, 'code.v'), code) or {
 		return 'Failed to write code to sandbox.'
 	}
-	build_res := isolate_cmd('isolate --box-id=$box_id --dir=$vexeroot --env=HOME=/box --processes=3 --mem=100000 --wall-time=2 --quota=${1048576 / block_size},${1048576 / inode_ratio} --run -- $vexeroot/v -cflags -DGC_MARKERS=1 -no-parallel -no-retry-compilation -g code.v')
```
Comment (spytheman) [inline server.v:99]: I think the compilation limit for the memory should be a separate constant, so that it can remain 100MB.

File: server.v [hunk 5]
```diff
@@ -81,28 +90,28 @@ fn ddhhmmss(time time.Time) string {
...
+	build_res := isolate_cmd('isolate --box-id=${box_id} --dir=${vexeroot} --env=HOME=/box --processes=${max_run_processes_and_threads} --mem=${max_run_memory_in_kb} --wall-time=${run_time_in_seconds} --run -- ${vexeroot}/v -cflags -DGC_MARKERS=1 -no-parallel -no-retry-compilation -g code.v')
 	build_output := build_res.output.trim_right('\n')
 	log_code(code, build_output) or { eprintln('[WARNING] Failed to log code.') }
 	if build_res.exit_code != 0 {
 		return prettify(build_output)
 	}
-	run_res := isolate_cmd('isolate --box-id=$box_id --dir=$vexeroot --env=HOME=/box --processes=1 --mem=30000 --wall-time=2 --quota=${10240 / block_size},${10240 / inode_ratio} --run -- code')
+	run_res := isolate_cmd('isolate --box-id=${box_id} --dir=${vexeroot} --env=HOME=/box --processes=${max_run_processes_and_threads} --mem=${max_run_memory_in_kb} --time=${run_time_in_seconds} --wall-time=${run_time_in_seconds} --run -- code')
```
Comment (spytheman) [inline server.v:114]: what is the difference between `--time` and `--wall-time` ?

File: server.v [hunk 6]
```diff
@@ -8,10 +8,17 @@ import sqlite
...
 const (
-	port        = 5555
-	vexeroot    = @VEXEROOT
-	block_size  = 4096
-	inode_ratio = 16384
+	port                          = 5555
+	vexeroot                      = @VEXEROOT
+	// non-standard block size, different for different filesystems
+	block_size                    = 4096
+	fs_usage_max_size_in_bytes    = 10 * 1024 * 1024
```
Comment (spytheman) [inline server.v:15]: The quota limit before was 1MB. Increasing it 10x may be too much.
How about `3 * 1024 * 1024` instead?

File: docker-compose.yml (hunk 1 above)
Comment (walkingdevel) [reply docker-compose.yml:6]: I don't know how the current production works, but without the privileged flag, cgroups used by the isolate doesn't work.

Comment (spytheman) [reply docker-compose.yml:6]: I think the Docker file is not used, it is just a convenience for developing the playground locally.

**Review by spytheman — APPROVED** [review]
Good work.
